### ISSUE TITLE
`error[E0786]: found invalid metadata files for crate` when building a rust host application (extism `0.2.0`)

### ISSUE THREAD

**Issue by k33g** [extism/extism#227]
Hello, I get this after running `cargo run`:

```bash
   Compiling log4rs v1.2.0
   Compiling wiggle v4.0.0
error[E0786]: found invalid metadata files for crate `file_per_thread_logger` which `wasmtime` depends on
  --> /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/wiggle-4.0.0/src/lib.rs:36:13
   |
36 |     pub use wasmtime::*;
   |             ^^^^^^^^
   |
   = note: failed to open rmeta metadata: '/home/ubuntu/samples/extism-xp/demo/target/debug/deps/libfile_per_thread_logger-81223c493d0f04ac.rmeta'
   = note: failed to open file '/home/ubuntu/samples/extism-xp/demo/target/debug/deps/libfile_per_thread_logger-81223c493d0f04ac.rlib': Operation not permitted (os error 1)

error: error writing dependencies to `/home/ubuntu/samples/extism-xp/demo/target/debug/deps/wiggle-c1426c5fb8af1d0f.d`: Operation not permitted (os error 1)

For more information about this error, try `rustc --explain E0786`.
error: could not compile `wiggle` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
```

**Comment by bhelx** [comment]
Hey @k33g, can you post your rustc version?

**Comment by bhelx** [comment]
Also can you specify what the project is? Is it a rust project using the rust SDK? running `cargo run` from where exactly?

**Comment by k33g** [comment]
👋 Hello @bhelx 

- `rustc 1.66.1 (90743e729 2023-01-10)`
- I use the Rust Host SDK with the sample of the doc https://extism.org/docs/integrate-into-your-codebase/rust-host-sdk

```bash
.
├── demo
│  ├── Cargo.lock
│  ├── Cargo.toml
│  ├── code.wasm
│  ├── src
│  │  └── main.rs
│  └── target
```
I run `cargo run` from the root of `demo` (`cd demo; cargo run`)

**Comment by nilslice** [comment]
@k33g - I have a feeling this is likely due to some lack of architecture support for your VM within the `wiggle` crate, or one of its dependencies. 

If you could, try to reduce the issue by creating a very simple project using `wiggle` alone: 

```sh
cargo new issue-reduction
cd issue-reduction
cargo add wiggle
cargo build 
```

If the same error message appears, we may need to take this to the `wiggle` authors, which happen to be the `wasmtime` team.

**Comment by bhelx** [comment]
yeah I can't seem to reproduce. On M1 (arm) Mac:

```bash
cargo new demo --bin
cd demo
echo 'extism = { git = "https://github.com/extism/extism.git", version = "0.2.0" }' >> Cargo.toml
cargo build # runs fine 
```

Do you need to pull it from git? Why not get it from the crates.io. Change your Cargo.toml:

```diff
-extism = { git = "https://github.com/extism/extism.git", version = "0.2.0" }
+extism = "0.2.0"
```

That might unblock you if you are stuck.

**Comment by k33g** [comment]
@nilslice @bhelx, the build was ok - I will re-create a VM and do all tests from the beginning

**Comment by k33g** [comment]
I did the same test with a new VM, more logs:
```bash
    Blocking waiting for file lock on build directory
   Compiling extism-runtime v0.2.0
   Compiling extism v0.2.0
error[E0786]: found invalid metadata files for crate `crypto_common` which `extism_runtime` depends on
 --> /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/extism-0.2.0/src/lib.rs:2:9
  |
2 | pub use extism_runtime::{
  |         ^^^^^^^^^^^^^^
  |
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libcrypto_common-f8cde36a6dd62a42.rmeta'
  = note: failed to open file '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libcrypto_common-f8cde36a6dd62a42.rlib': Operation not permitted (os error 1)
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libcrypto_common-9876bd7ced42c215.rmeta'

error[E0786]: found invalid metadata files for crate `digest` which `extism_runtime` depends on
 --> /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/extism-0.2.0/src/lib.rs:2:9
  |
2 | pub use extism_runtime::{
  |         ^^^^^^^^^^^^^^
  |
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rmeta'
  = note: failed to open file '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rlib': Operation not permitted (os error 1)
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c7a59f6be345859.rmeta'

error[E0786]: found invalid metadata files for crate `sha2` which `extism_runtime` depends on
 --> /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/extism-0.2.0/src/lib.rs:2:9
  |
2 | pub use extism_runtime::{
  |         ^^^^^^^^^^^^^^
  |
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libsha2-0537ce112573444d.rmeta'
  = note: failed to open file '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libsha2-0537ce112573444d.rlib': Operation not permitted (os error 1)
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libsha2-82f7887b1a67d9ca.rmeta'

error[E0786]: found invalid metadata files for crate `digest` which `extism_runtime` depends on
 --> /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/extism-0.2.0/src/context.rs:3:63
  |
3 | pub struct Context(pub(crate) std::sync::Arc<std::sync::Mutex<extism_runtime::Context>>);
  |                                                               ^^^^^^^^^^^^^^
  |
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rmeta'
  = note: failed to open file '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rlib': Operation not permitted (os error 1)
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c7a59f6be345859.rmeta'

error[E0786]: found invalid metadata files for crate `digest` which `extism_runtime` depends on
  --> /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/extism-0.2.0/src/context.rs:15:13
   |
15 |             extism_runtime::Context::new(),
   |             ^^^^^^^^^^^^^^
   |
   = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rmeta'
   = note: failed to open file '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rlib': Operation not permitted (os error 1)
   = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c7a59f6be345859.rmeta'

error[E0786]: found invalid metadata files for crate `digest` which `extism_runtime` depends on
  --> /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/extism-0.2.0/src/context.rs:24:56
   |
24 |     pub(crate) fn lock(&self) -> std::sync::MutexGuard<extism_runtime::Context> {
   |                                                        ^^^^^^^^^^^^^^
   |
   = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rmeta'
   = note: failed to open file '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rlib': Operation not permitted (os error 1)
   = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c7a59f6be345859.rmeta'

error[E0786]: found invalid metadata files for crate `digest` which `extism_runtime` depends on
 --> /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/extism-0.2.0/src/plugin.rs:5:9
  |
5 |     id: extism_runtime::PluginIndex,
  |         ^^^^^^^^^^^^^^
  |
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rmeta'
  = note: failed to open file '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c74de6b03f8bfb7.rlib': Operation not permitted (os error 1)
  = note: failed to open rmeta metadata: '/home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c7a59f6be345859.rmeta'

error[E0460]: found possibly newer version of crate `digest` which `extism_runtime` depends on
  --> /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/extism-0.2.0/src/plugin.rs:26:49
   |
26 |         functions: impl IntoIterator<Item = &'a extism_runtime::Function>,
   |                                                 ^^^^^^^^^^^^^^
   |
   = note: perhaps that crate needs to be recompiled?
   = note: the following crate versions were found:
           crate `digest`: /home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libdigest-3c7a59f6be345859.rmeta
           crate `extism_runtime`: /home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/libextism_runtime-2be858830823a3f7.rmeta

error: could not write output to /home/ubuntu/samples/rust-experiments/hello-world/target/debug/deps/extism_runtime-2be858830823a3f7.extism_runtime.e0d3a70c-cgu.2.rcgu.o: Operation not permitted

error: could not compile `extism-runtime` due to previous error
warning: build failed, waiting for other jobs to finish...
For more information about this error, try `rustc --explain E0786`.
error: could not compile `extism` due to 8 previous errors
```
I will try to do the same on a PI4 (8Go RAM)

**Comment by bhelx** [comment]
@k33g i'm gonna assume you got this working because we watching your demo live at Wasm I/O :)

feel free to open it up again if you are stuck though. Be sure to upgrade to 0.4.0 while you are at it.

**Comment by nilslice** [comment]
@liudonghua123 - what version of Extism are you running, and can you share a bit about your environment / OS / architecture?

**Comment by liudonghua123** [comment]
> @liudonghua123 - what version of Extism are you running, and can you share a bit about your environment / OS / architecture?

I'm not running Extism, I encountered the similar error when building a tauri demo app.
